python/ray/ml/result.py

```diff
@@ -4,15 +4,31 @@
 
 
 class Result:
-    """The final result of a ML training run.
+    """The final result of a ML training run or a Tune trial.
 
-    This is the class produced by Trainer.fit(). It contains a checkpoint, which
-    can be used for resuming training and for creating a Predictor object. It also
-    contains a metrics object describing training metrics.
+    This is the class produced by Trainer.fit() or Tuner.fit() (through ResultGrid).
+    It contains a checkpoint, which can be used for resuming training and for
+    creating a Predictor object. It also contains a metrics object describing
+    training metrics. `status` and `error` are included so that non successful runs
+    and trials can be represented as well.
+
+    The constructor is a private API.
+
+    Args:
+        metrics: The final metrics as reported by an Trainable.
+        checkpoint: The final checkpoint of the Trainable
+        error: The execution error of the Trainable run, if the trial finishes in error.
```

Comment:
```suggestion
    Attributes:
        metrics (Any): The final metrics as reported by an Trainable.
        checkpoint (Optional[Checkpoint]): The final checkpoint of the Trainable
        error (Optional[Exception]): The execution error of the Trainable run, if the trial finishes in error.
```

FYI, we should include typing in the docstring here because they are publicly accessible **attributes** to the user.

Note that this is different from the args passed in to the constructor, which we wouldn't include typing here, since the constructor is private.

Reply:
Do you mean specifically for attributes? The Ray code style guide says to *not* show types for args fields.

Reply:
Can we clarify what the guideline is? Right now, I am referring to the style here: https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/ml/trainer.py?L31

Reply:
It should follow this style: https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#code-style

(no types in pydoc)

It should indeed say Attributes instead of Args though. Args would be for the pydoc of the init function only.

Reply:
Got it, updated to "attributes"